README.md
# PingStatus [WIP]
This is a small IoT device that pings a set of IP addresses and lights up LEDs depending on their status.  
It can be used to check if a service or website is reachable, or if a device has entered or left the local network.  
The firmware is still unstable and might not work at times. Please check back later for an updated version.

<br>

## Features:
- Observes any amount of clients by pinging them via ICMP (private / public IP) or an HTTP request (URL)
- Automatically dims LEDs using PWM based on the ambient light level for optimal brightness level
- Automatically shuts off LEDs for even less light pollution at night
  - Momentary button to temporarily disable auto-shutoff mode
- Internal clock is updated with NTP, so it is always consistant and relatively accurate
- Many configuration settings for setting it up to exactly suit your needs

<br>

---

<br>

## Table of Contents:

- [Features](#features)
- [Parts list](#parts-list)
- [Setup](#setup)
- [Build](#build)
    - [Build Guide](#build-guide)
    - [Circuit Schematic](#circuit-schematic)
    - [Enclosure 3D Model](#enclosure-3d-model)

<br>

---

<br><br>

## Parts list:
### Required parts:
> - ESP board (ESP-8266 for example)
>     - Note: it needs to have at least 1 analog pin and 1 digital pin per observed client, plus 1 for the optional push button
> - Pre-perforated circuit board and a few wires for bridging
> - Micro USB cable and USB power supply

<br>

### Parts needed per observed client:
> - LED of any color
> - ~150 Ω resistor
>     - Value depends on the forward voltage of the LED but 150 Ω *should* fit most

<br>

### Optional parts:
> - Light Dependent Resistor / Photoresistor (LDR)
>     - Wavelength: optimally around 480-540 nm - Resistance: doesn't really matter as a multiplier can be configured
> - 10 kΩ pull-down resistor for LDR  
>   
> These parts can be skipped if you disable auto dimming in the config  

<br>

> - Momentary push button or limit switch (normally open)
> - 10 kΩ pull-down resistor for push button  
>   
> These parts can also be skipped if you don't want to be able to temporarily disable the auto-shutoff mode

<br><br>

## Setup:
1. Install the [Arduino IDE](https://www.arduino.cc/en/software)
2. Add your ESP board's URL in the IDE preferences by following [this guide](https://arduino-esp8266.readthedocs.io/en/latest/installing.html)
3. Install your ESP board's firmware in the board manager
4. Install the following libraries by downloading them and extracting them to `<Arduino folder>/libraries` (usually under "Documents"):
    - [ESP8266Ping](https://github.com/dancol90/ESP8266Ping/archive/refs/heads/master.zip)
    - [ArduinoJson](https://github.com/bblanchon/ArduinoJson/archive/refs/heads/master.zip)
5. [Download the PingStatus repository](https://github.com/Sv443/PingStatus/archive/refs/heads/master.zip) and extract it somewhere
6. Rename the file `config.h.template` to `config.h`
7. Edit `config.h` to provide your WiFi settings and configure everything else to your needs
8. Open the .ino file with the Arduino IDE
9. Connect your ESP board to your computer via USB and wait for the drivers to be installed and ready
10. Match up the board settings under `Tools` with your ESP board and select the right COM or ttyUSB port
11. Hit the upload button
12. Open the serial monitor with <kbd>Ctrl</kbd> <kbd>Shift</kbd> <kbd>M</kbd> to verify that it works
    - If the output is garbled, reset the board. If it is still unreadable, make sure the BAUD rate is set to 115200.

<br><br>

## Build:
### Build Guide:
just connect the thingies to the thingies lol

<br>

### Circuit Schematic:
[Schematic coming soon]

<br>

### Enclosure 3D Model:
[Model coming soon]

<br><br><br>

<div align="center" style="text-align:center;">

Made with ❤️ by [Sv443](https://sv443.net/)  
If you like PingStatus, please consider [supporting the development](https://github.com/sponsors/Sv443)

</div>


CHANGELOG.md
## Roadmap
- Add support for GET/HEAD requests
- Option to run a webhook when a client goes down (and back up)
- Option to accumulate multiple clients' ping results into a single LED

<br>

## v0.2.0 (WIP)
- LEDs are now auto-dimmed based on amount of ambient light by reading an LDR
- Added DNS resolution support to send a ping request to the IP behind a provided domain
- Added optional push button to manually refresh the clients or temporarily disable the night mode
- Changed client configuration to a JSON string for easier modification and more flexibility

<br>

## v0.1.0
- Added basic pinging of any amount of clients
- Added auto-dimmer function
- Added night mode


dev/PingStatus.old.cpp
#include "config.hpp"
#include "types.hpp"


/** Raw time object */
time_t now;
/** More convenient time wrapper */
tm tm;

void setup() {
  #if SERIAL_ENABLED
    Serial.begin(SERIAL_BAUD);

    Serial.println();
    Serial.println("PingStatus by Sv443 (https://github.com/Sv443)");
    Serial.println();

    #if PING_INTERVAL < 30000
      Serial.println("!!");
      Serial.println("!! Warning: Setting PING_INTERVAL below ~30s might cause erratic or delayed behavior if the network and processing latency is too high");
      Serial.println("!!");
    #endif
  #endif

  #if AUTO_DIMMING_ENABLED
    configTime(TIME_ZONE, NTP_SERVER);
  #endif

  WiFi.persistent(false);
  WiFi.mode(WIFI_STA);
  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);

  #if SERIAL_ENABLED
    Serial.print("WiFi connecting");
  #endif

  #if STATUS_LED_ENABLED
    pinMode(STATUS_LED_PIN, OUTPUT);
  #endif
  int ledState = ACTIVE_LOW ? HIGH : LOW;

  while(WiFi.status() != WL_CONNECTED) {
    #if SERIAL_ENABLED
      Serial.print(".");
    #endif

    #if STATUS_LED_ENABLED
      digitalWrite(STATUS_LED_PIN, ledState);
      ledState = ledState == HIGH ? LOW : HIGH;
    #endif

    delay(500);
  }

  digitalWrite(STATUS_LED_PIN, ACTIVE_LOW ? LOW : HIGH);

  #if SERIAL_ENABLED
    Serial.println();
  #endif

  for (int i = 0; i < CLIENTS_AMOUNT; i++)
    pinMode(clients[i].pin, OUTPUT);
}

/** Updates the globals `now` and `tm` with the current time */
void updateTime() {
  time(&now);
  localtime_r(&now, &tm);
}

void loop() {
  for(int i = 0; i < CLIENTS_AMOUNT; i++) {
    IPAddress ip = clients[i].ip;
    int pin = clients[i].pin;

  #if SERIAL_ENABLED
    Serial.print("\nPinging client #");
    Serial.print(i + 1);
    Serial.print(" (IP: ");
    Serial.print(ip);
    Serial.print(" - Pin: ");
    Serial.print(pin);
    Serial.println(")");
  #endif

    bool pingRes = Ping.ping(ip);

    if(pingRes) {
      #if SERIAL_ENABLED
        Serial.print("  > Client #");
        Serial.print(i + 1);
        Serial.println(" is available\n");
      #endif
    }
    else {
      #if SERIAL_ENABLED
        Serial.print("  > Client #");
        Serial.print(i + 1);
        Serial.println(" is not available\n");
      #endif
    }

    // TODO: collect LED states in a variable, then update at the end to update all at once
    updateLedState(pin, pingRes);
  }

  // TODO: change to like 100ms, then calculate amount of interval with PING_INTERVAL
  delay(PING_INTERVAL);
}

/** Updates the LED at `pin` to be `enabled` or not, using the optional brightness and dim values set in config.h */
void updateLedState(int pin, bool enabled) {
  #if AUTO_DIMMING_ENABLED
    updateTime();

    #if SERIAL_ENABLED
      Serial.print("Current time: ");
      Serial.print(tm.tm_hour);
      Serial.print(":");
      Serial.print(tm.tm_min);
      Serial.print(":");
      Serial.println(tm.tm_sec);
    #endif

    // TODO:
    // 1. read LDR analog value
    // 2. offset / multiply by value from config.h
    // 3. clamp between 0 and 10 to reduce resolution and "jitter"
    // 4. set LED brightness

    bool ledsOn = betweenThreshold(ledsOnThreshold, ledsOffThreshold);
    bool bright = true; // TODO

    if(!ledsOn) {
      analogWrite(pin, 0);

      // TODO: status led blink sequence to notify of the LEDs turning off
      // maybe the LEDs should blink too?
      #if SERIAL_ENABLED && SERIAL_DEBUG
        Serial.print("Setting LED at pin ");
        Serial.print(pin);
        Serial.println(" to disabled (off period)");
      #endif

      return;
    }

    #if SERIAL_ENABLED && SERIAL_DEBUG
      Serial.print("Setting LED at pin ");
      Serial.print(pin);
      Serial.print(" to enabled ");
      Serial.println(bright ? "(on & bright period)" : "(on & dim period)");
    #endif
    if(bright) {
      int val = enabled ? floor(LED_BRIGHT_VALUE * 1024) : 0;
      analogWrite(pin, val);
    }
    else {
      int val = enabled ? floor(LED_DIM_VALUE * 1024) : 0;
      Serial.println(val);
      analogWrite(pin, val);
    }
  #else
    Serial.print("Setting LED at pin ");
    Serial.print(pin);
    Serial.print(" to ");

    if(!ledsOn) {
      digitalWrite(pin, LOW);

      // TODO: status led blink sequence to notify of the LEDs turning off
      // maybe the LEDs should blink too?
      #if SERIAL_ENABLED && SERIAL_DEBUG
        Serial.println("disabled (off period)");
      #endif
    }
    else {
      digitalWrite(pin, enabled);

      #if SERIAL_ENABLED && SERIAL_DEBUG
        Serial.print(enabled ? "enabled" : "disabled");
        Serial.println(" (on period)");
      #endif
    }
  #endif
}

/** Checks if the global `tm` is between the `min` and `max` thresholds in minutes */
bool timeBetweenThreshold(uint16 min, uint16 max) {
  return tm.tm_hour * 60 + tm.tm_min >= min
    && tm.tm_hour * 60 + tm.tm_min < max;
}

/** Parses an IPv4 string to an IPAddress instance */
IPAddress parseIP(String ip) {
  // TODO: parse IP
  return IPAddress(1, 2, 3, 4);
}


dev/config.template.old.hpp
#pragma once
#include "types.h"

//     ____             __ _
//    / ___|___  _ __  / _(_) __ _ _ 
//   | |   / _ \| '_ \| |_| |/ _` (_)
//   | |__| (_) | | | |  _| | (_| |_ 
//    \____\___/|_| |_|_| |_|\__, (_)
//                           |___/
//
// This file contains your WiFi password!
// Make sure to treat it accordingly!


/*****************************/
/*           Time            */
/*****************************/

// Your time zone - possible values: https://github.com/esp8266/Arduino/blob/master/cores/esp8266/TZ.h
#define TIME_ZONE TZ_Etc_UTC          // UTC
//#define TIME_ZONE TZ_Europe_Berlin  // Europe, Berlin
//#define TIME_ZONE TZ_Etc_GMTp1      // GMT+1
//#define TIME_ZONE TZ_Etc_GMTm3      // GMT-3


/*****************************/
/*           WiFi            */
/*****************************/

// WiFi access point SSID (name)
#define WIFI_SSID ""

// WiFi access point password
#define WIFI_PASSWORD ""


/*****************************/
/*          Clients          */
/*****************************/

// Amount of clients that are being observed
#define CLIENTS_AMOUNT 2

// Enter all client IP addresses you'd like to observe and the GPIO pin of their LED
//   Note: The GPIO and digital pinouts are different! Check your board's pinout diagram to see which is which
//   If you want to make use of the auto-dimming, make sure the pins support PWM (squiggly line on your board's diagram)
//   If some of the pins don't support PWM, make sure to set AUTO_DIMMING_ENABLED to false
//   After changing this, please adjust CLIENTS_AMOUNT above
//
// Format:
//   ClientType  type               // Can be ClientType.IP or ClientType.URL
//   String      ipOrUrl            // IP address or URL
//   int         pin                // GPIO pin of the LED
//   int         ldrValueBounds[2]  // Lower and upper bounds of the LDR's value (can be between 0 and 1023) (only used if AUTO_DIMMING_ENABLED is true)
//   bool        activeLow          // If your LEDs are active low, set this to true, else false
//
PingClient clients[CLIENTS_AMOUNT] = {
  {
    ClientType.IP,
    "192.168.1.1",
    5,
    {0, 500},
    false,
  },
};

// Interval in milliseconds between sending pings (default = 10s/client)
//   Note: Anything under (10 seconds * amount of clients) might cause erratic or delayed behavior if the network and processing latency is too high
#define PING_INTERVAL 10000 * CLIENTS_AMOUNT


/*****************************/
/*       Auto-Dimming        */
/*****************************/

// Whether auto-dimming should be enabled
//   Requires all LEDs to be hooked up to PWM pins (squiggly line on your board's diagram)
#define AUTO_DIMMING_ENABLED true

// The maximum possible brightness of the LEDs as a value between 0.0 and 1.0
#define LED_MAX_VALUE 1.0
// The lowest possible brightness of the LEDs as a value between 0.0 and 1.0
#define LED_MIN_VALUE 0.025


// If this is enabled, the LEDs will be turned off according to the settings below
#define LED_OFF_ENABLED true

// When to turn the LEDs on (24 hours)
#define LED_ON_HOUR 6
#define LED_ON_MIN  0

// When to turn the LEDs off (24 hours)
#define LED_OFF_HOUR 24
#define LED_OFF_MIN  0


// NTP server pool URL
#define NTP_SERVER "de.europe.pool.ntp.org"
//#define NTP_SERVER "pool.ntp.org"           // Default (uses the closest NTP server)
//#define NTP_SERVER "europe.pool.ntp.org"    // European NTP server pool
//#define NTP_SERVER "us.pool.ntp.org"        // US NTP server pool


/*****************************/
/*           Other           */
/*****************************/

// Whether the status LED (WiFi connection & activity indicator) should be enabled
#define STATUS_LED_ENABLED true

// Whether the status LED is active low (true on ESP, false on Arduino)
#define STATUS_LED_ACTIVE_LOW true

// Set to a GPIO pin if you want an external status LED
#define STATUS_LED_PIN BUILTIN_LED

// Whether the serial monitor should be enabled
#define SERIAL_ENABLED true

// BAUD rate of the serial monitor
#define SERIAL_BAUD 115200

// Enable to get more debug output in the serial monitor
#define SERIAL_DEBUG true


config.template.hpp
#pragma once
#include "types.hpp"

//     ____             __ _
//    / ___|___  _ __  / _(_) __ _ _ 
//   | |   / _ \| '_ \| |_| |/ _` (_)
//   | |__| (_) | | | |  _| | (_| |_ 
//    \____\___/|_| |_|_| |_|\__, (_)
//                           |___/
//
// This file contains your WiFi password!
// Make sure to treat it accordingly!


/*************************************/
/*               Pins                */
/*************************************/

// GPIO pin of the manual refresh button (-1 to disable)
//   Can be any digital pin
#define PIN_BUTTON -1

// GPIO pin of the light dependent resistor (LDR)
//   This has to be an analog pin!
#define PIN_LDR -1


/*************************************/
/*               WiFi                */
/*************************************/

// WiFi access point SSID (name)
#define WIFI_SSID ""

// WiFi access point password
#define WIFI_PASSWORD ""


/*************************************/
/*               Time                */
/*************************************/

// Your time zone - possible values: https://github.com/esp8266/Arduino/blob/master/cores/esp8266/TZ.h
//#define TIME_ZONE TZ_Etc_UTC        // UTC
#define TIME_ZONE TZ_Europe_Berlin    // Europe, Berlin
//#define TIME_ZONE TZ_Etc_GMTp5      // GMT+5
//#define TIME_ZONE TZ_Etc_GMTm3      // GMT-3

// URL or IP to an NTP (time) server
//#define NTP_SERVER "pool.ntp.org"           // Default (uses the closest NTP server)
#define NTP_SERVER "192.168.178.1"            // Use a specific IP (like your router) as the NTP server
//#define NTP_SERVER "europe.pool.ntp.org"    // Closest NTP server in Europe
//#define NTP_SERVER "us.pool.ntp.org"        // Closest NTP server in the US


/*************************************/
/*              Clients              */
/*************************************/

// Timeout in milliseconds between refreshing the status of every client (default = 1 minute)
#define REFRESH_TIMEOUT 1000 * 60

// The clients to ping
//   TODO: explain JSON structure
String clients = R"([
  {
    "name": "my cool phone",
    "ip": "192.168.420.69",
    "pin": 5,
    "ldrMultiplier": 1.0,
    "activeLow": false,
  },
  {
    "name": "my cool website",
    "domain": "www.sv443.net",
    "pin": 14,
    "ldrMultiplier": 0.5,
    "activeLow": false,
  }
])";


/*************************************/
/*          Serial Monitor           */
/*************************************/

// Whether the default USB serial monitor should be enabled
#define SERIAL_ENABLED true

// BAUD rate of the serial monitor
#define SERIAL_BAUD 115200

// This defines the amount of log messages you'll receive in the serial monitor
//   Possible values:
//   LogLevel.Debug   - Prints pretty much everything
//   LogLevel.Info    - Prints only the most important info (default option)
//   LogLevel.Warning - Only prints warnings and errors
#define SERIAL_LOG_LEVEL LogLevel.Info


/*************************************/
/*            Night Mode             */
/*************************************/

// If this is enabled, the LEDs will be turned off at night according to the settings below
#define NIGHT_MODE_ENABLED true

// Define the lower and upper boundary of the LDR (aka what values are considered the darkest and lightest)
//   The value has to be between 0 and 1023
//   If your SERIAL_LOG_LEVEL is set to LogLevel.Debug the serial monitor will print out the LDR's current value
#define LDR_MIN_VAL 0
#define LDR_MAX_VAL 500

// When to turn the LEDs off (24 hour format, local time)
#define NM_OFF_HOUR 23
#define NM_OFF_MIN  0

// When to turn the LEDs back on again (24 hour format, local time)
#define NM_ON_HOUR 6
#define NM_ON_MIN  0


/*************************************/
/*            Status LED             */
/*************************************/

// Whether the status LED (WiFi connection & activity indicator) should be enabled
#define STATUS_LED_ENABLED true

// Whether the status LED is active low (true on ESP builtin, false on Arduino builtin)
#define STATUS_LED_ACTIVE_LOW true

// Replace with a GPIO pin if you want an external status LED
#define STATUS_LED_PIN BUILTIN_LED
